File: day2/main.cpp
#include <iostream>
#include <string>
#include <cstdint>
#include <vector>
#include <sstream>

bool is_invalid(uint64_t number) {
    std::string number_str = std::to_string(number);
    if (number_str.size() % 2 == 0 &&
        number_str.substr(0, number_str.size() / 2) == 
        number_str.substr(number_str.size() / 2)) {
        return true;
    }
    return false;
}

bool is_invalid_b(uint64_t number) {
    std::string number_str = std::to_string(number);
    for (uint32_t i = 2; i <= number_str.size(); i++) {
        if (number_str.size() % i != 0) {
            continue;
        }
        uint32_t seq_size = number_str.size() / i;
        bool is_invalid = true;
        for (uint32_t j = 0; j < i - 1; j++) {
            if (number_str.substr(j * seq_size, seq_size) !=
                number_str.substr(j * seq_size + seq_size, seq_size)) {
                is_invalid = false;
                break;
            }
        }
        if (is_invalid) {
            return true;
        }
    }
    return false;
}

int main() {
    std::string input;

    std::cin >> input;

    std::vector<std::pair<uint64_t, uint64_t>> ranges;
    std::stringstream ss(input);
    uint64_t res_a = 0;
    uint64_t res_b = 0;

    while (std::getline(ss, input, ',')) {
        std::stringstream ss2(input);
        uint64_t start, end;
        ss2 >> start;
        ss2.ignore(1);
        ss2 >> end;
        ranges.push_back(std::make_pair(start, end));
    }

    for (const auto& range : ranges) {
        for (uint64_t i = range.first; i <= range.second; i++) {
            if (is_invalid(i)) {
                res_a += i;
            }
            if (is_invalid_b(i)) {
                res_b += i;
            }
        }
    }

    std::cout << res_a << std::endl;
    std::cout << res_b << std::endl;
    return 0;
}

File: day2/main.rs
use std::io;
use std::str::FromStr;

fn is_invalid(val : u64) -> bool {
    let val_str = val.to_string();
    
    if val_str.len() % 2 == 0 {
        let mid = val_str.len() / 2;
        let (left, right) = val_str.split_at(mid);
        return left == right;
    }

    return false;
}

fn is_invalid_b(val: u64) -> bool {
    let val_str = val.to_string();

    for i in 2..=val_str.len() {
        if val_str.len() % i != 0 {
            continue;
        }
        let seq_size = val_str.len() / i;
        let mut is_invalid = true;
        for j in 0..i-1 {
            let seq1 = &val_str[j * seq_size..(j + 1) * seq_size];
            let seq2 = &val_str[(j + 1) * seq_size..(j + 2) * seq_size];
            if seq1 != seq2 {
                is_invalid = false;
                break;
            }
        }
        if is_invalid {
            return true;
        }
    }

    return false;
}

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let ranges: Vec<&str> = input.split(',').collect();

    let mut res_a: u64 = 0;
    let mut res_b: u64 = 0;
    for range in ranges {
        let mut parts = range.split('-');
        let a = parts.next().unwrap();
        let b = parts.next().unwrap();
        let a_val = u64::from_str(&a).unwrap();
        let b_val = u64::from_str(&b).unwrap();

        for val in a_val..=b_val {
            if is_invalid(val) {
                res_a += val;
            }
            if is_invalid_b(val) {
                res_b += val;
            }
        }
    }

    println!("{res_a}");
    println!("{res_b}");
}